sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala

```diff
@@ -0,0 +1,271 @@
+    Arguments:
+      * input      - string value to mask. Supported types: STRING, VARCHAR, CHAR
+      * upperChar  - character to replace upper-case characters with. Specify -1 to retain original character. Default value: 'X'
+      * lowerChar  - character to replace lower-case characters with. Specify -1 to retain original character. Default value: 'x'
+      * digitChar  - character to replace digit characters with. Specify -1 to retain original character. Default value: 'n'
+      * otherChar  - character to replace all other characters with. Specify -1 to retain original character. Default value: -1
+  """,
+  examples = """
+    Examples:
+      > SELECT _FUNC_('abcd-EFGH-8765-4321');
+        xxxx-XXXX-nnnn-nnnn
+      > SELECT _FUNC_('abcd-EFGH-8765-4321', 'Q');
+        xxxx-QQQQ-nnnn-nnnn
+      > SELECT _FUNC_('AbCD123-@$#', 'Q', 'q');
+        QqQQnnn-@$#
+      > SELECT _FUNC_('AbCD123-@$#');
+        XxXXnnn-@$#
+      > SELECT _FUNC_('AbCD123-@$#', 'Q');
+        QxQQnnn-@$#
+      > SELECT _FUNC_('AbCD123-@$#', 'Q', 'q');
+        QqQQnnn-@$#
+      > SELECT _FUNC_('AbCD123-@$#', 'Q', 'q', 'd');
+        QqQQddd-@$#
+      > SELECT _FUNC_('AbCD123-@$#', 'Q', 'q', 'd', 'o');
+        QqQQdddoooo
+      > SELECT _FUNC_('AbCD123-@$#', -1, 'q', 'd', 'o');
+        AqCDdddoooo
+      > SELECT _FUNC_('AbCD123-@$#', -1, -1, 'd', 'o');
+        AbCDdddoooo
+      > SELECT _FUNC_('AbCD123-@$#', -1, -1, -1, 'o');
+        AbCD123oooo
+      > SELECT _FUNC_(NULL, -1, -1, -1, 'o');
+        NULL
+      > SELECT _FUNC_(NULL);
+        NULL
+      > SELECT _FUNC_('AbCD123-@$#', -1, -1, -1, -1);
+        AbCD123-@$#
+  """,
+  since = "3.4.0",
+  group = "string_funcs")
+// scalastyle:on line.size.limit
+case class Mask(
+    input: Expression,
+    upperChar: Expression,
+    lowerChar: Expression,
+    digitChar: Expression,
+    otherChar: Expression)
+    extends QuinaryExpression
+    with ImplicitCastInputTypes
+    with QueryErrorsBase
+    with NullIntolerant {
+
+  def this(input: Expression) =
+    this(
+      input,
+      Literal(Mask.MASKED_UPPERCASE),
+      Literal(Mask.MASKED_LOWERCASE),
+      Literal(Mask.MASKED_DIGIT),
+      Literal(Mask.MASKED_IGNORE))
+
+  def this(input: Expression, upperChar: Expression) =
+    this(
+      input,
+      upperChar,
+      Literal(Mask.MASKED_LOWERCASE),
+      Literal(Mask.MASKED_DIGIT),
+      Literal(Mask.MASKED_IGNORE))
+
+  def this(input: Expression, upperChar: Expression, lowerChar: Expression) =
+    this(input, upperChar, lowerChar, Literal(Mask.MASKED_DIGIT), Literal(Mask.MASKED_IGNORE))
+
+  def this(
+      input: Expression,
+      upperChar: Expression,
+      lowerChar: Expression,
+      digitChar: Expression) =
+    this(input, upperChar, lowerChar, digitChar, Literal(Mask.MASKED_IGNORE))
+
+  override def checkInputDataTypes(): TypeCheckResult = {
+
+    def checkInputDataType(exp: Expression, message: String): Option[TypeCheckResult] = {
+      if (!exp.foldable) {
```

Comment:
Do we need to check the length of the replacing string as 1?

Reply:
Yes, added a check for replacing string size a 1